### PR TITLE
dnsproxy: introduce cache to reference count regexes

### PR DESCRIPTION
**Pull request by odinuge**
This introduce a reference counted cache/index for regexes, and is ready for supporting other caches as well. This means that policies with the same set of rules can reuse the underlying matcher saving, since long regex tends to occupy a lot of memory.

This also makes the LRU cache for compiled regexes more or less redundant, mitigating the drawbacks of that cache. Having a large LRU cache is useful when a lot of similar policies are in use at the same time, but in case a node have a small set of policies inserted at the same time, but the churn makes new and distinct policies come and go, the LRU cache will use a lot more memory than the benefit it gives.

benchmark TL;DR: Heap usage compared to a LRU size of 128 is down by ~99%, and ~92% with a LRU size of 1024, with the given test data*.

* defining heap usage as (HeapInuse after test) - (HeapInuse before setup)
```
$ go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x ./pkg/fqdn/dnsproxy

// With this change;

go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x -memprofile memprofile.out ./pkg/fqdn/dnsproxy goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=128)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 20 MiB    NumGC = 6
Before Test (N=1,EPs=20,cache=128)
Alloc = 37 MiB  HeapInuse = 49 MiB      Sys = 1031 MiB  NumGC = 18
After Test (N=1,EPs=20,cache=128)
Alloc = 52 MiB  HeapInuse = 62 MiB      Sys = 1031 MiB  NumGC = 53
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        4289436000 ns/op        2559832616 B/op 34303565 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      12.574s

// Baseline with LRU size of 1024

goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=1024)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 20 MiB    NumGC = 6
Before Test (N=1,EPs=20,cache=1024)
Alloc = 37 MiB  HeapInuse = 49 MiB      Sys = 969 MiB   NumGC = 18
After Test (N=1,EPs=20,cache=1024)
Alloc = 312 MiB HeapInuse = 357 MiB     Sys = 969 MiB   NumGC = 46
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        5990697400 ns/op        3470756528 B/op 36068464 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      11.813s

// Baseline with LRU size of 128
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=128)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 19 MiB    NumGC = 5
Before Test (N=1,EPs=20,cache=128)
Alloc = 38 MiB  HeapInuse = 49 MiB      Sys = 1047 MiB  NumGC = 17
After Test (N=1,EPs=20,cache=128)
Alloc = 1763 MiB        HeapInuse = 1907 MiB    Sys = 2918 MiB  NumGC = 44
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        81058653300 ns/op       10328100808 B/op        46896658 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      93.574s
```

```release-note
Introduce smarter internal cache to reduce memory consumption for FQDN / DNS policy usage, especially in environment with heavy FQDN / DNS policy usage
```

### CODE REVIEW

**Comment by odinuge** [comment]
A step towards https://github.com/cilium/cilium/pull/20396 and https://github.com/cilium/cilium/pull/20246, that is smaller and easier to reason around.

**Comment by odinuge** [comment]
cc @christarazi

File: pkg/fqdn/dnsproxy/proxy.go [hunk 1]
```diff
@@ -157,22 +161,77 @@ type DNSProxy struct {
...
+// fqdnMatcher is used to match an FQDN to see if the ruleset allows it or not.
+// In case a matcher is nil, it means that the respective matcher doesn't match any FQDNs
+type fqdnMatcher struct {
+	// patternMatcher is used to match an FQDN to the matchPatterns from a DNS policy
+	patternMatcher *regexp.Regexp
+}
+
+// CachedSelectorEntry maps port numbers to selectors to rules, mirroring
+// policy.L7DataMap but the DNS rules are compiled into a regex
+type CachedSelectorEntry map[policy.CachedSelector]fqdnMatcher
```
Comment (odinuge) [inline pkg/fqdn/dnsproxy/proxy.go]: We can flattern this right now to be `*regexp.Regexp` instead of `fqdnMatcher`. If/when we want the map/list of matchNames separately, we can introduce the `fqdnMatcher`

File: pkg/fqdn/dnsproxy/proxy.go [hunk 2]
```diff
@@ -157,22 +161,77 @@ type DNSProxy struct {
...
+	}
+}
+
+// matchFQDN checks if the given fqdn is matched by the underlying policy, using the matcher
+func (r fqdnMatcher) matchFQDN(fqdn string) bool {
+	if r.patternMatcher != nil && r.patternMatcher.MatchString(fqdn) {
+		return true
+	}
+	return false
+}
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]:
```suggestion
	return r.patternMatcher != nil && r.patternMatcher.MatchString(fqdn)
}
```

File: pkg/fqdn/dnsproxy/proxy.go (hunk 1 above)
Comment (christarazi) [reply pkg/fqdn/dnsproxy/proxy.go]: Sure, let's flatten it now.

File: pkg/fqdn/dnsproxy/proxy.go [hunk 3]
```diff
@@ -274,61 +363,154 @@ func (p *DNSProxy) RemoveRestoredRules(endpointID uint16) {
 	p.removeRestoredRulesLocked(uint64(endpointID))
 }
 
+// lookupOrCompilePatternMatcher will check if the pattern is already compiled and present in another policy, and
+// will reuse it in order to reduce memory consumption. The usage is reference counted, so all calls where
+// lookupOrCompilePatternMatcher returns no error, a subsequent call to release it via releasePatternMatcher has to
+// be tone when its no longer being used by the policy.
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]:
```suggestion
// be done when it's no longer being used by the policy.
```

File: pkg/fqdn/dnsproxy/proxy.go [hunk 4]
```diff
@@ -274,61 +363,154 @@ func (p *DNSProxy) RemoveRestoredRules(endpointID uint16) {
...
+// releasePatternMatcher releases the pattern matcher. In case there are no longer any references to it,
+// it will be freed. Running release on a nil value is a noop.
+func (c *matcherCache) releasePatternMatcher(patternMatcher *regexp.Regexp) {
+	if patternMatcher == nil {
+		return
+	}
+	if indexEntry, ok := c.patternMatcherCache[patternMatcher.String()]; ok {
+		switch indexEntry.referenceCount {
+		case 1:
+			delete(c.patternMatcherCache, patternMatcher.String())
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]: Let's call `String()` only once so we don't waste cycles doing duplicate work

File: pkg/fqdn/restore/restore.go [hunk 5]
```diff
@@ -21,20 +20,24 @@ type IPRules []IPRule
 
 // IPRule stores the allowed destination IPs for a DNS names matching a regex
 type IPRule struct {
-	Re  RuleRegex
+	// v1.12 and older versions depend on having this value provided as a string, but for future
+	// improvements we gracefully handle reading it as nil as well. We always ensure that we currently
+	// never write nil in order to break rolling back to v1.12 or earlier.
+	Re  *string
```
Comment (christarazi) [inline pkg/fqdn/restore/restore.go]: Have you tested downgrading from this version and validated that the rules are restored properly and traffic is not disrupted?

**Review by christarazi — CHANGES_REQUESTED** [review]
Looks great! A few comments

**Comment by odinuge** [comment]
Hey! Thanks for the review and the comments! I've been caught up with other work the last two weeks while being on-call, but have/will set aside as much time as needed to get this over the line next week starting on Tuesday.

Thanks
Odin

File: pkg/fqdn/dnsproxy/proxy.go (hunk 2 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy.go]: Thanks! Ended up refactoring this out since I removed that `fqdnMatcher` struct

File: pkg/fqdn/dnsproxy/proxy.go (hunk 3 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy.go]: Good catch, thanks!

File: pkg/fqdn/dnsproxy/proxy.go (hunk 4 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy.go]: Yeah, sounds good!

File: pkg/fqdn/restore/restore.go (hunk 5 above)
Comment (odinuge) [reply pkg/fqdn/restore/restore.go]: Ended up "reverting" this change and keeping "99%" of the old logic here to ensure we behave the same way.

The actual disk format is exactly the same, so are the patterns/regexes.

I guess we can write some more tests, tho. I think it should be ok covered by `TestRestoredEndpoint` and `TestFullPathDependence`.

**Comment by odinuge** [comment]
👋 Pushed a few changes now based on your comments Chris. Pushed them as `--fixup` commits so that its easier to track the diff, and then I'll squash them when things are ready!

I'll give it another look tomorrow morning, and maybe add a few more tests of the edge-cases.

Thanks
Odin

File: pkg/fqdn/restore/restore.go [hunk 6]
```diff
@@ -25,17 +24,24 @@ type IPRule struct {
 	IPs map[string]struct{} // IPs, nil set is wildcard and allows all IPs!
 }
 
-// RuleRegex is a wrapper for *regexp.Regexp so that we can define marshalers for it.
+// RuleRegex is a wrapper for a pointer to a string so that we can define marshalers for it.
 type RuleRegex struct {
-	*regexp.Regexp
+	Pattern *string
 }
```
Comment (christarazi) [inline pkg/fqdn/restore/restore.go:30]: I feel like this change along with the changes to the marshaling function can be in their own separate commit. Even though it helps with the overall optimization of introducing a cache, it can still be standalone because it sort of an optimization itself + at the same time being a preparatory commit for the main cache optimization.

File: pkg/fqdn/dnsproxy/proxy.go [hunk 7]
```diff
@@ -249,7 +284,27 @@ func (p *DNSProxy) RestoreRules(ep *endpoint.Endpoint) {
...
+	restoredRules := make(map[uint16][]restoredIPRule, len(ep.DNSRules))
+	for port, dnsRule := range ep.DNSRules {
+		ipRules := make([]restoredIPRule, 0, len(dnsRule))
+		for _, ipRule := range dnsRule {
+			var regex *regexp.Regexp
+			if ipRule.Re.Pattern != nil {
+				var err error
+				regex, err = p.cache.lookupOrCompileRegex(*ipRule.Re.Pattern)
+				if err != nil {
+					log.Warnf("Error compiling regex for endpointID %d, gracefully ignoring", ep.ID)
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy.go]: I think `INFO` level is more appropriate for this log msg. I think we'll need a more user-friendly msg as well.

```suggestion
					log.WithFields(logrus.Fields{
					  logfields.EndpointID: ep.ID,
					  logfields.Rule: *ipRule.Re.Pattern,
					}).Info("Disregarding restored DNS rule due to failure in compiling regex. Traffic to the FQDN may be disrupted.")
```

File: pkg/fqdn/restore/restore.go (hunk 6 above)
Comment (odinuge) [reply pkg/fqdn/restore/restore.go:30]: Yee, agree! Will split them!

File: pkg/fqdn/dnsproxy/proxy.go [hunk 8]
```diff
@@ -210,8 +214,11 @@ func (p *DNSProxy) checkRestored(endpointID uint64, destPort uint16, destIP stri
 	}
 
 	for i := range ipRules {
-		if _, exists := ipRules[i].IPs[destIP]; (exists || ipRules[i].IPs == nil) && ipRules[i].Re.MatchString(name) {
-			return true
+		ipRule := ipRules[i]
+		if _, exists := ipRule.IPs[destIP]; exists || ipRule.IPs == nil {
+			if ipRule.regex != nil && ipRule.regex.MatchString(name) {
```
Comment (odinuge) [inline pkg/fqdn/dnsproxy/proxy.go:220]: Should I skip this `nil` check so we can still keep it "undefined" until we decide on the actual semantics of a `nil` value? Tho. I think treating `nil` as no match is probably the most meaningful one.

File: pkg/fqdn/dnsproxy/proxy.go [hunk 9]
```diff
@@ -249,7 +284,30 @@ func (p *DNSProxy) RestoreRules(ep *endpoint.Endpoint) {
 	if ep.IPv6.IsValid() {
 		p.restoredEPs[ep.IPv6.String()] = ep
 	}
-	p.restored[uint64(ep.ID)] = ep.DNSRules
+	restoredRules := make(map[uint16][]restoredIPRule, len(ep.DNSRules))
+	for port, dnsRule := range ep.DNSRules {
+		ipRules := make([]restoredIPRule, 0, len(dnsRule))
+		for _, ipRule := range dnsRule {
+			if ipRule.Re.Pattern == nil {
```
Comment (odinuge) [inline pkg/fqdn/dnsproxy/proxy.go:292]: This is also assuming we define a `null` value as not allowed

File: pkg/fqdn/dnsproxy/proxy_test.go [hunk 10]
```diff
@@ -17,6 +17,8 @@ import (
 	"testing"
 	"time"
 
+	"golang.org/x/exp/maps"
+
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy_test.go]: Nit: no need for extra newline

File: pkg/fqdn/dnsproxy/proxy_test.go [hunk 11]
```diff
@@ -1000,6 +1002,35 @@ func (s *DNSProxyTestSuite) TestRestoredEndpoint(c *C) {
 	_, exists = s.proxy.restored[epID1]
 	c.Assert(exists, Equals, false)
 
+	invalidRePattern := "invalid-re-pattern((*"
+	validRePattern := "^this[.]domain[.]com[.]$"
+	k := maps.Keys(restored)[0]
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy_test.go]: `maps.Keys()` returns the keys in an indeterminate order. Does the test make any assmptions about the order of the keys returned? 

```
// Keys returns the keys of the map m.
// The keys will be in an indeterminate order.
```

https://github.com/odinuge/cilium/blob/e28cce148c676371a234a51f303c8c7933f81960/vendor/golang.org/x/exp/maps/maps.go#L10

File: pkg/fqdn/dnsproxy/proxy.go (hunk 8 above)
Comment (christarazi) [reply pkg/fqdn/dnsproxy/proxy.go:220]: Is it possible for it to be `nil` or is this just coming from overly defensive coding? In the end, I think treating the `nil` as no match makes sense regardless.

Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy.go:220]: Hmm, yeah, mostly defensive since we discussed this in a previous PR in regards to the list of `matchNames`.

File: pkg/fqdn/dnsproxy/proxy_test.go (hunk 10 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy_test.go]: Ahh, goimports moved there... I'll move it down and reformat! Thanks

File: pkg/fqdn/dnsproxy/proxy_test.go (hunk 11 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy_test.go]: The order doesn't matter since this map only has one element.

I _think_ the key is actually random since its the port of the dns-server, and it listens to `:0` that I think means "any free" port or so. Guess we could parse `s.proxy.TCPServer.Listener.Addr()` as well, but I thought this was the simplest. Maybe I should add a comment about it

ref. https://github.com/cilium/cilium/blob/e28cce148c676371a234a51f303c8c7933f81960/pkg/fqdn/dnsproxy/proxy_test.go#L120

File: pkg/fqdn/dnsproxy/proxy.go (hunk 8 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy.go:220]: > Is it possible for it to be nil

as far as i understand, the only reason for nil is if someone manually change the json file on disk to either nil or `""` (I think), so it should never happen

Comment (christarazi) [reply pkg/fqdn/dnsproxy/proxy.go:220]: Might as well keep it. We can treat it as no-match. In case someone accidentally mucks with the json file or somehow the data is corrupted, we don't want to inadvertently allow traffic, even though the chances of this happening are extremely small.

**Comment by christarazi** [comment]
@odinuge This seems good to go in my book for wider review. If you feel it's ready, please mark it ready for review by moving it out of draft. Thanks!

**Comment by odinuge** [comment]
> This seems good to go in my book for wider review. If you feel it's ready, please mark it ready for review by moving it out of draft. Thanks!

Thanks for all your thorough(and very fast) reviews @christarazi! I'll give it a last look either later this weekend or Monday morning to check if I have missed anything obvious, and will mark as ready to review after that!

**Comment by odinuge** [comment]
I did find another edge case btw. where `len(l7Rules.DNS) == 0` i changed the behavior to block all traffic instead of allowing all traffic as the baseline/master does. Pushed this diff to keep the current behavior from master; https://github.com/cilium/cilium/compare/492e9cc5c08f0e9fca8ccfe6ae762c13bbbd1284..d6a38a28c9176b5d0e769b33ef5e250107f86713

That case ^ results in an empty regex (the pattern is an empty string), and an empty regex match all input (the parsed syntax is actually `(?:)` in go). That is possibly another optimization we can do to check if the pattern is empty, and in that case, define it as a "always allow" rule we don't need to run `MatchString` on; but thats for another PR.

File: pkg/fqdn/dnsproxy/proxy_test.go [hunk 12]
```diff
@@ -437,6 +449,32 @@ func (s *DNSProxyTestSuite) TestRespondMixedCaseInRequestResponse(c *C) {
...
+func (s *DNSProxyTestSuite) TestCheckNoRules(c *C) {
+	name := "cilium.io."
+	l7map := policy.L7DataMap{
+		cachedDstID1Selector: &policy.PerSelectorPolicy{
+			L7Rules: api.L7Rules{},
+		},
+	}
+	query := name
+
+	err := s.proxy.UpdateAllowed(epID1, dstPort, l7map)
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy_test.go:462]: Lint check failed here

File: pkg/fqdn/dnsproxy/proxy_test.go [hunk 13]
```diff
@@ -437,6 +449,32 @@ func (s *DNSProxyTestSuite) TestRespondMixedCaseInRequestResponse(c *C) {
...
+	c.Assert(allowed, Equals, true, Commentf("request was rejected when it should be allowed"))
+
+	l7map = policy.L7DataMap{
+		cachedDstID1Selector: &policy.PerSelectorPolicy{
+			L7Rules: api.L7Rules{
+				DNS: []api.PortRuleDNS{},
+			},
+		},
+	}
+	err = s.proxy.UpdateAllowed(epID1, dstPort, l7map)
```
Comment (christarazi) [inline pkg/fqdn/dnsproxy/proxy_test.go:477]: Lint check failed here as well

**Comment by christarazi** [comment]
@odinuge It would be good to fix the above issues & rebase on latest master, so that we can run the full CI suite.

File: pkg/fqdn/dnsproxy/proxy_test.go (hunk 13 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy_test.go:477]: Ooof, will fix, thanks!

**Comment by odinuge** [comment]
Hmm. Both failing tests seem like unrelated flakes to me. Mind triggering the tests again?

**Comment by christarazi** [comment]
From what I can tell, yes these are flakes. Tracking issues:

* https://github.com/cilium/cilium/issues/21710
* https://github.com/cilium/cilium/issues/16203

**Comment by ti-mo** [comment]
Note: the privileged_tests build tag has been removed in favor of the PRIVILEGED_TESTS env, in case you run into any surprises after rebase.

**Comment by christarazi** [comment]
/test-1.16-4.9

**Comment by odinuge** [comment]
Hi @christarazi @ti-mo! Are you waiting for me here, or? Happy to engage if there is anything you want me to do! Would be great to get an agreement here so we could start pushing the actual regex improvements to reduce the heap size and allocations/cpu usage during compilation. 😄

**Review by christarazi — APPROVED** [review]

**Comment by christarazi** [comment]
@odinuge Given that this is a user-facing change, could you write the release note from the user's point of view and the sort of impact that they will see? I'm thinking something like "reduced memory usage with FQDN policies" but with more elaboration.

**Comment by odinuge** [comment]
👋  @christarazi!

Anything you want me to do here?

**Comment by christarazi** [comment]
@odinuge I don't think there's much left to do here. I think the code looks good. You have tested & validated it your environments which is great. Pending another CI pass, I think it should be ready to merge.

One point for follow up is where does this leave us with the FQDN LRU? Could we simply deprecate it and remove it in the next cycle (1.14)?

**Comment by odinuge** [comment]
> @odinuge I don't think there's much left to do here. I think the code looks good. You have tested & validated it your environments which is great. Pending another CI pass, I think it should be ready to merge.

Great!

> One point for follow up is where does this leave us with the FQDN LRU? Could we simply deprecate it and remove it in the next cycle (1.14)?

Yeah, that is a good question. I vote for at least removing the lru from the dnsproxy since we now (with this) have a separate cache; but let's do it in a follow up after this.

The other places it might be useful, since we reduce the actual time spent during regex compilation; but haven't digged too deep into those yet.

One thing I do vote for is to remove [this](https://github.com/cilium/cilium/blob/master/pkg/policy/api/fqdn.go#L91) and [this](https://github.com/cilium/cilium/blob/master/pkg/policy/api/fqdn.go#L123) since we already validate that the `matchPattern` is valid. Unless there is some combination of characters that can lead to regex compilation failing when it matches the `allowedPatternChars` regex above?

**Comment by christarazi** [comment]
* 5.4 hit https://github.com/cilium/cilium/issues/21735. 
* External workloads failing due to provisioning GKE cluster.

Given these are flakes, I think this is ready to merge.
